streaming/receiver.c

```diff
@@ -455,12 +453,20 @@ static int rrdpush_receive(struct receiver_state *rpt)
     if (!netdata_exit && rpt->host) {
         netdata_mutex_lock(&rpt->host->receiver_lock);
         if (rpt->host->receiver == rpt) {
+            netdata_mutex_unlock(&rpt->host->receiver_lock);
```

Comment:
Backing out of the lock is dangerous as it allows the receiver to be detached from the host by another thread. Why not take the host lock over the whole if-nest L453-472?

Reply:
Today's lock order is 😜 :
- `host_lock`
- `global_lock`
- `receiver_lock`

Reply:
Yes, I figured that was the approach. I agree with the locking order shown in the PR:
1. Global
2. Host
3. Receiver

From memory I think this is consistent with the code-base but we need to check.  I can see a problem with this - the flags are protected under the host lock which is root cause of needing to back-out / relock in several places. I don't want to fix that in this PR but we should put a new flags lock in place *only* to cover the flags. I did something similar in the dead branch for the replication code. Then I think we should use:

1. Global
2. Host
3. Receiver
...
n. Flags

i.e. it should always be the last lock in the order.